### PR TITLE
`svm_state::AccountStorage`: phase out `svm-storage` and `svm-kv`

### PR DESCRIPTION
**Pull request by neysofu**
This PR introduces an `AccountStorage` utility, which allows to read/write data related to an `Account`. In particular, raw variable segmentation (using 32 byte segments currently) logic which is then used for smWasm host calls `get64`, `get32`, `set64`, and `set32`.

Closes https://github.com/spacemeshos/svm/issues/358.

### CODE REVIEW

File: crates/state/src/global_state.rs [hunk 1]
```diff
@@ -1,20 +1,24 @@
...
 /// querying and other features which make it suitable for storing Spacemesh'
 /// global state.
 ///
 /// This data structure is backed by SQLite.
+#[derive(Debug, Clone)]
 pub struct GlobalState {
-    pub(crate) storage: Storage,
+    pub(crate) storage: Arc<Mutex<Storage>>,
+    runtime: Arc<Mutex<Runtime>>,
+    historical_query_context: Option<Layer>,
```
Comment (YaronWittenstein) [inline crates/state/src/global_state.rs:21]: this name feels a bit too wordy - is it like the `HEAD` reference used in Git? (metaphorically)

File: crates/state/src/global_state.rs [hunk 2]
```diff
@@ -24,264 +28,137 @@ impl GlobalState {
...
 
-    pub async fn checkpoint(&mut self) -> Result<()> {
-        self.storage.checkpoint().await?;
+    /// Saves dirty changes in preparation of [`GlobalState::commit`]. After
+    /// saving, changes are frozen and can't be removed from the current layer.
+    ///
+    /// This might return a [`StorageError::KeyCollision`] depending on the
+    /// content of the dirty changes, so beware.
+    pub fn checkpoint(&mut self) -> Result<()> {
+        self.block_on(self.storage().checkpoint())?;
```
Comment (YaronWittenstein) [inline crates/state/src/global_state.rs:65]: not critical but I'd personally decompose into: (to reduce some cognitive load)

```rust
let storage = self.storage();
self.block_on(storage.checkpoint())?
```

File: crates/state/src/global_state.rs [hunk 3]
```diff
@@ -24,264 +28,137 @@ impl GlobalState {
...
 
-    pub async fn commit(&mut self) -> Result<(Layer, Fingerprint)> {
-        let res = self.storage.commit().await?;
-        Ok(res)
+    /// Persists all changes to disk and returns the root [`State`] of the new
+    /// layer. It returns a [`StorageError::DirtyChanges`] in case there's any
+    /// dirty changes that haven't been saved via [`GlobalState::checkpoint`]
+    /// before this call.
+    pub fn commit(&mut self) -> Result<(Layer, State)> {
+        Ok(self.block_on(self.storage().commit())?)
```
Comment (YaronWittenstein) [inline crates/state/src/global_state.rs:74]: same comment as the last one.

File: crates/state/src/global_state.rs [hunk 4]
```diff
@@ -24,264 +28,137 @@ impl GlobalState {
...
     }
 
-    pub async fn current_layer(&mut self) -> Result<(Layer, Fingerprint)> {
-        let res = self.storage.last_layer().await?;
-        Ok(res)
+    /// Returns the [`Layer`] and [`State`] of the last ever committed
+    /// layer; i.e. persisted changes without dirty and saved changes.
+    pub fn current_layer(&mut self) -> Result<(Layer, State)> {
+        Ok(self.block_on(self.storage().last_layer())?)
```
Comment (YaronWittenstein) [inline crates/state/src/global_state.rs:80]: same comment as the last one.

File: crates/state/src/global_state.rs [hunk 5]
```diff
@@ -24,264 +28,137 @@ impl GlobalState {
...
     }
 
-    pub async fn rollback(&mut self) -> Result<()> {
-        self.storage.rollback().await?;
+    /// Erases all dirty changes from memory. Persisted and saved data are left
+    /// untouched.
+    pub fn rollback(&mut self) -> Result<()> {
+        self.block_on(self.storage().rollback())?;
```
Comment (YaronWittenstein) [inline crates/state/src/global_state.rs:86]: same comment as the last one.

File: crates/state/src/global_state.rs [hunk 6]
```diff
@@ -24,264 +28,137 @@ impl GlobalState {
...
+    /// after and excluding `layer_id` from the SQLite store. Persisted data is
+    /// left untouched. It returns a [`StorageError::DirtyChanges`] in case
+    /// there's any dirty changes, i.e. you must call [`GlobalState::rollback`]
+    /// beforehand.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `layer_id` is invalid.
+    pub fn rewind(&mut self, layer_id: Layer) -> Result<()> {
+        self.block_on(self.storage().rewind(layer_id))?;
```
Comment (YaronWittenstein) [inline crates/state/src/global_state.rs:100]: same comment as the last one.

File: crates/state/src/template_storage.rs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    pub fn new(template_addr: &TemplateAddr, gs: GlobalState) -> Self {
+        Self {
+            template_addr: template_addr.clone(),
+            gs,
+        }
+    }
+
+    /// Reads, decodes and finally returns all [`Sections`] of `self`.
+    pub fn sections(&self) -> StorageResult<Option<Sections>> {
+        let core_sections_opt: Option<Sections> =
```
Comment (YaronWittenstein) [inline crates/state/src/template_storage.rs:24]: can you please omit the `_opt` suffix?

File: crates/state/src/storage.rs [hunk 8]
```diff
@@ -175,7 +175,7 @@ impl Storage {
 
     async fn get_by_hash_historical(
         &self,
-        hash: &Fingerprint,
+        hash: &State,
         layer_id: Layer,
     ) -> Result<Option<Vec<u8>>> {
         let value_opt: Option<(Vec<u8>,)> = sqlx::query_as(
```
Comment (YaronWittenstein) [inline crates/state/src/storage.rs:181]: can you please omit the `_opt` suffix?

**Review by YaronWittenstein — CHANGES_REQUESTED** [review]
Amazing job!!

I've left a few comments - we're almost there!

File: crates/runtime/src/vmcalls/storage.rs [hunk 9]
```diff
@@ -1,60 +1,29 @@
...
-}
-
 /// Stores memory cells `[mem_ptr, mem_ptr + 1, ..., mem_ptr + 19]` into variable `var_id`.
 ///
 /// # Panics
 ///
 /// Panics if variable `var_id`'s length isn't 20 bytes.
 pub fn store160(env: &FuncEnv, mem_ptr: u32, var_id: u32) {
-    store_n_impl!(20, env, mem_ptr, var_id);
+    let bytes: [u8; 20] = {
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/runtime/src/vmcalls/storage.rs]: Done!

File: crates/runtime/src/vmcalls/storage.rs [hunk 10]
```diff
@@ -65,62 +34,50 @@ pub fn store160(env: &FuncEnv, mem_ptr: u32, var_id: u32) {
 ///
 /// Panics if variable `var_id`'s length isn't 20 bytes.
 pub fn load160(env: &FuncEnv, var_id: u32, mem_ptr: u32) {
-    load_n_impl!(20, env, var_id, mem_ptr);
+    let borrow = env.borrow();
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/runtime/src/vmcalls/storage.rs]: Done!

File: crates/state/src/account_storage.rs [hunk 11]
```diff
@@ -1,92 +1,512 @@
...
-        let slice = &self.var(var_id).await[offset..offset + byte_size];
-        i64::from_be_bytes(slice.try_into().unwrap())
+impl Segment {
+    fn len(&self) -> usize {
+        // E.g. `0..=2` contains 3 elements.
+        *self.range.end() - *self.range.start() + 1
     }
+}
+
+fn var_segments(account_addr: &Address, offset: u32, byte_size: u32) -> Vec<Segment> {
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/account_storage.rs:272]: Gotcha. This makes ton of sense. Implementing variable segmentation logic is quite complex though - speaking from experience... I'll create a new issue and deal with this in a new PR.

File: crates/state/src/account_storage.rs [hunk 12]
```diff
@@ -1,92 +1,512 @@
...
+            template_addr,
+        })
     }
+}
 
-    pub async fn set_var_160(&mut self, var_id: u32, value: [u8; 20]) {
-        debug_assert_eq!(self.layout.get(Id(var_id)).byte_size(), 20);
+pub struct AccountMut {
+    pub balance: u64,
+    pub counter: u64,
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/account_storage.rs:347]: What's wrong with two `u64` values?

File: crates/state/src/account_storage.rs [hunk 13]
```diff
@@ -1,92 +1,512 @@
...
-        let mut value = self.var(var_id).await;
-        let key = keys::account_var(&self.address, var_id, &self.layout);
-        let offset = self.layout.get(Id(var_id)).offset() as usize % 32;
+impl Codec for AccountData {
+    type Error = ParseError;
 
-        let slice = &mut value[offset..offset + new_value.len()];
-        slice.copy_from_slice(new_value);
+    fn encode(&self, w: &mut impl WriteExt) {
+        let encoding_version = 0u8;
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/account_storage.rs:321]: :+1:

File: crates/state/src/account_storage.rs [hunk 14]
```diff
@@ -1,92 +1,512 @@
...
+
     use super::*;
 
-    pub fn account_var(account_addr: &Address, var_id: u32, layout: &FixedLayout) -> String {
-        let offset = layout.get(Id(var_id)).offset();
-        let key_index = offset % 32;
+    fn fixed_layout() -> FixedLayout {
+        let mut builder = FixedLayoutBuilder::new();
+
+        builder.set_first(Id(1));
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/account_storage.rs:390]: Much better now, thanks.

File: crates/state/src/account_storage.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/state/src/account_storage.rs:347]: I didn't explain myself (my bad) - I've meant to make the `counter` a `u128` (supporting any possible Nonce Scheme we'd like to implement). The `balance` should indeed be `u64`

File: crates/state/src/account_storage.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/state/src/account_storage.rs:272]: yes - simple is better...

File: crates/state/src/account_storage.rs [hunk 15]
```diff
@@ -1,92 +1,512 @@
...
+        let template_addr = TemplateAddr::repeat(0x80);
+        let name = "@name";
+        let balance = 42;
+        let counter = 0;
+
+        let gs = GlobalState::in_memory();
+        let mut account = AccountStorage::new(gs, &address, &template_addr, &layout);
+        account.create(name.to_string(), template_addr, balance, counter);
+
+        assert_eq!(account.name(&address).unwrap().unwrap(), name);
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/account_storage.rs]: `AccountData` and `AccountMut` are not part of the public interaface, I'd rather test the public endpoints (e.g. `.name()` and `.balance()`). What do you think?

File: crates/state/src/global_state.rs (hunk 2 above)
Comment (neysofu) [reply crates/state/src/global_state.rs:65]: I'd prefer to leave these as simple one-liners if that's okay with you! I don't think there's much added value in `let storage = self.storage()`.

File: crates/state/src/global_state.rs (hunk 1 above)
Comment (neysofu) [reply crates/state/src/global_state.rs:21]: Yes. I agree it's way too verbose, but this is complex logic and I opted for the more descriptive name. Any tips?

File: crates/state/src/error.rs [hunk 16]
```diff
@@ -18,16 +14,23 @@ pub enum StorageError {
     #[error("Key collision from two different checkpoints.")]
     KeyCollision {
         /// They Blake3 hash of the key that caused this collision.
-        key_hash: Fingerprint,
+        key_hash: [u8; 32],
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/error.rs:17]: I realized that `Fingerprint` was basically a duplicate of `svm_types::State`, so I decided to remove it altogether. And while `[u8; 32]` is the same as a `State`, semantically they're two different things (one is an obscure state signature for accounts, GS, templates, etc., while `key_hash` is... well, a hash).

File: crates/state/src/template_storage.rs (hunk 7 above)
Comment (neysofu) [reply crates/state/src/template_storage.rs]: Done, it's better now. Sorry about doing this again, I lost this change after the last review.

File: crates/state/src/storage.rs (hunk 8 above)
Comment (neysofu) [reply crates/state/src/storage.rs:181]: Done.

File: crates/state/src/global_state.rs (hunk 2 above)
Comment (neysofu) [reply crates/state/src/global_state.rs:65]: I've kept this one open and marked as Resolved all the others.

File: crates/state/src/error.rs [hunk 17]
```diff
@@ -18,16 +14,23 @@ pub enum StorageError {
     #[error("Key collision from two different checkpoints.")]
     KeyCollision {
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/error.rs:15]: Exactly (not SHA-3 though, Blake3). Replacing a variable with the same hash is obviously allowed, but not within the same checkpoint. When a value gets overwritten or created twice within a single checkpoint, we'll raise a `KeyCollision` error.

[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/error.rs:15]: I'll wait for your feedback before marking this as Resolved.

File: crates/state/src/account_storage.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/account_storage.rs]: Done. `u128` is also now supported by the `Codec` trait!

File: crates/state/src/account_storage.rs [hunk 18]
```diff
@@ -1,92 +1,512 @@
...
+    fn encode(&self, w: &mut impl WriteExt) {
+        let encoding_version = 0u8;
+
+        encoding_version.encode(w);
+        self.balance.encode(w);
+        self.counter.encode(w);
+    }
+
+    fn decode(reader: &mut impl ReadExt) -> std::result::Result<Self, Self::Error> {
+        let encoding_version = u8::decode(reader)?;
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/account_storage.rs:374]: Done.

File: crates/state/src/global_state.rs (hunk 2 above)
Comment (YaronWittenstein) [reply crates/state/src/global_state.rs:65]: ok, it's up to you (it's a personal taste matter)

File: crates/state/src/global_state.rs (hunk 1 above)
Comment (YaronWittenstein) [reply crates/state/src/global_state.rs:21]: let's open an issue for that in order to avoid blocking this PR

File: crates/runtime/Cargo.toml [hunk 19]
```diff
@@ -21,9 +22,7 @@ wat = "1"
 svm-hash = { path = "../hash" }
 svm-types = { path = "../types" }
 svm-layout = { path = "../layout" }
-svm-kv = { path = "../kv", default-features = false }
 svm-state = { path = "../state" }
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/runtime/Cargo.toml:25]: `svm-state` supports both by default! There's no need to add feature flags (`GlobalState::new` and `GlobalState::in_memory`, respectively, do what the features did). What indeed we must do, though, is choosing one or the other initializer based on `svm-runtime` features.

File: crates/state/src/error.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/state/src/error.rs:17]: Yes, just thought it would be better to avoid repetitive `[u8; 32]` in the code

[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/error.rs:17]: Let's use `State` then, I'm fine with it. I'll change this.

File: crates/state/src/global_state.rs (hunk 1 above)
Comment (neysofu) [reply crates/state/src/global_state.rs:21]: https://github.com/spacemeshos/svm/issues/368

File: crates/state/src/error.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/state/src/error.rs]: As agreed on Slack, this will be removed. We don't want to trigger an error in this situation.
